### PR TITLE
Connect swaps/swap claims

### PR DESCRIPTION
**Pull request by jessepinho**
Despite the `SwapView` and `SwapClaimView` types supporting fields that allow swaps and swap claims to reference each other, we haven't actually been populating those fields yet.

This PR adds a couple keys to the `TransactionPerspective` Protobuf type to make it possible to populate the `claim_tx` and `swap_tx` fields in `SwapView` and `SwapClaimView`, respectively. Hopefully, in the future, we can do the same with undelegates <-> undelegate claims.

### CODE REVIEW

File: proto/penumbra/penumbra/core/transaction/v1/transaction.proto
Comment (jessepinho) [inline proto/penumbra/penumbra/core/transaction/v1/transaction.proto:122]: Reviewers, please let me know what you think of this approach. It's the only way I could think of to easily allow views with links to other transactions to be generated from perspectives.

In the future, we'll also need to figure out a way to stuff output 1 and output 2 from the swap claim into the `TransactionPerspective` of a swap. But this work has already been taking me quite a while, so I wanted to at least get this merged first.

File: crates/core/transaction/src/view/transaction_perspective.rs
Comment (redshiftzero) [inline crates/core/transaction/src/view/transaction_perspective.rs:45]: should these fields also be getting populated in `ViewServer::transaction_info_by_hash` in the `penumbra-view` crate?

Comment (jessepinho) [reply crates/core/transaction/src/view/transaction_perspective.rs:45]: Probably, but I figured I'd leave that up to core devs. These fields aren't a hard requirement — they're just _available_ (after this PR is merged) to use for generating richer views. We'll be populated these fields in the web app, and now the core devs can too if they choose to.

Comment (hdevalence) [reply crates/core/transaction/src/view/transaction_perspective.rs:45]: Yeah, I think it's fine for the Rust view server to lag a bit, until there's a concrete use case (like a web frontend pulling from it).

File: crates/core/component/dex/src/swap/view.rs
Comment (hdevalence) [inline crates/core/component/dex/src/swap/view.rs:82]: Don't we want this metadata too? We'd need it to render the assets (symbols, icons, etc) for a `Swap` without a corresponding `SwapClaim` right?

File: proto/penumbra/penumbra/core/transaction/v1/transaction.proto
Comment (hdevalence) [reply proto/penumbra/penumbra/core/transaction/v1/transaction.proto:122]: Two comments:

1. I think the comment should read "transactions creating the commitments that they nullify" -- this is a minor nit but it's more technically correct since a swap commitment is a different kind of commitment than an output note

2. Why do we do base64 encoding of the nullifiers or commitments? This would force all of the consumers to add in special base64 parsing to interpret the perspective, right? Could we instead encode a list of `(Nullifier, TransactionId)` or `(Commitment, TransactionId)` pairs?

File: crates/core/transaction/src/is_action.rs
Comment (hdevalence) [inline crates/core/transaction/src/is_action.rs:366]: Does this work? I'd be surprised if `nullifier.to_string()` returned a Base64 string. There's in general no guarantee that the `ToString` impl on the Rust domain type has any relation to the protojson encoding. For instance `Address`'s `ToString` does bech32.

File: proto/penumbra/penumbra/core/transaction/v1/transaction.proto
Comment (hdevalence) [reply proto/penumbra/penumbra/core/transaction/v1/transaction.proto:122]: I think there are restrictions on what types Proto3 allows to be used in maps -- like they have to be scalar types (?) -- so this would mean defining two sub-messages, each modeling one of the pairs, and then having a `repeated PairType` (which is basically the way that a `map` will be encoded on the wire anyways). Then we'd change the domain type to have a `BTreeMap<A, B>` and deserialize the proto-gen-type's `Vec<PairType>` into `BTreeMap<A, B>` using `.collect`.

File: crates/core/transaction/src/is_action.rs
Comment (hdevalence) [inline crates/core/transaction/src/is_action.rs:348]: The output notes of the swap are determined _entirely_ by (1) the `SwapPlaintext` (committed to by the swap commitment and (2) the `BatchSwapOutputData`.  This means that to get the output notes, we should supply a BSOD in the perspective, then recompute them.

Compared to trying to insert the output notes manually, this has two advantages:

1. It correctly handles the case where the swap tx has been accepted on chain but the claim has not yet been submitted. In that case, even though we don't know the swap claim tx (it doesn't exist), we can still show the output notes.
2. It allows rendering extra information (if we want), like the % share of the batch the user's swap was.

File: crates/core/transaction/src/is_action.rs
Comment (hdevalence) [inline crates/core/transaction/src/is_action.rs:367]: Populating the `asset_1_metadata` and `asset_2_metadata` is important to be able to view the `Swap` action pre-submission, where there won't be any output notes.

File: proto/penumbra/penumbra/core/transaction/v1/transaction.proto
Comment (hdevalence) [inline proto/penumbra/penumbra/core/transaction/v1/transaction.proto:151]: This needs to be `repeated` since there could be multiple different `Swap`s in the same tx.

**Review by hdevalence — APPROVED** [review]

**Comment by hdevalence** [comment]
Rebased on main to resolve conflicts in the generated proto descriptors.

File: crates/core/transaction/src/view/transaction_perspective.rs
Comment (hdevalence) [reply crates/core/transaction/src/view/transaction_perspective.rs]: Filed #4083

File: crates/core/component/dex/src/swap/view.rs
Comment (hdevalence) [reply crates/core/component/dex/src/swap/view.rs]: xref https://github.com/penumbra-zone/penumbra/pull/4035#discussion_r1535134495

**Comment by grod220** [comment]
Nice job! This was trickier than I expected.

**Pull request merged**
